File: RaftSwift/Sources/RaftCore/Node/Persistence/InMemoryRaftNodePersistence.swift
/// In memory implementation of node persistence.
///
/// This implementation of persistence is not persistent accos node failures, since the snapshots are saved in memory.
public actor InMemoryRaftNodePersistence: RaftNodePersistence {
    var snapshots: [Int: Snapshot] = [:]

    public let compactionThreshold: Int

    public init(compactionThreshold: Int) {
        self.compactionThreshold = compactionThreshold
    }

    public func saveSnapshot(_ snapshot: Snapshot, for nodeId: Int) async throws {
        snapshots[nodeId] = snapshot
    }

    public func loadSnapshot(for nodeId: Int) async throws -> Snapshot? {
        snapshots[nodeId]
    }

    public func deleteSnapshot(for nodeId: Int) async throws {
        snapshots[nodeId] = nil
    }
}

File: RaftKotlin/raft-node/src/main/kotlin/raft/core/node/persistence/InMemoryRaftNodePersistence.kt
package raft.core.node.persistence

import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import raft.core.node.RaftNodePersistence
import raft.core.utils.types.Snapshot

/**
 * In memory implementation of node persistence.
 *
 * This implementation of persistence is not persistent across node failures,
 * since the snapshots are saved in memory.
 */
class InMemoryRaftNodePersistence(
    override val compactionThreshold: Int
) : RaftNodePersistence {

    private val snapshots: MutableMap<Int, Snapshot> = mutableMapOf()
    private val mutex = Mutex()

    override suspend fun saveSnapshot(snapshot: Snapshot, nodeId: Int) {
        mutex.withLock {
            snapshots[nodeId] = snapshot
        }
    }

    override suspend fun loadSnapshot(nodeId: Int): Snapshot? {
        mutex.withLock {
            return snapshots[nodeId]
        }
    }

    override suspend fun deleteSnapshot(nodeId: Int) {
        mutex.withLock {
            snapshots.remove(nodeId)
        }
    }
}